pd/src/state/writer.rs

```diff
@@ -259,6 +289,75 @@ impl Writer {
             )
             .execute(&mut dbtx)
             .await?;
+
+            // If the note was previously quarantined, drop it from quarantine
+            query!(
+                "DELETE FROM quarantined_notes WHERE note_commitment = $1",
+                &<[u8; 32]>::from(note_commitment)[..]
+            )
+            .execute(&mut dbtx)
+            .await?;
+        }
+
+        // Calculate the height at which notes quarantined in this block should unbond. If the
+        // unbonding period or the epoch duration change, notes will unbond at the nearest epoch
+        // boundary following this height.
+        let unbonding_epochs = self
+            .private_reader()
+            .chain_params_rx()
+            .borrow()
+            .unbonding_epochs;
+        let epoch_duration = block.epoch.as_ref().unwrap().duration;
+        let unbonding_height = height + (epoch_duration * unbonding_epochs);
```

Comment:
We calculate _at the time of undelegation submission_ the height at which the notes should become available (modulo slashing) and store _that_ height in the database. This naturally means that if the chain's epoch duration or unbonding epochs are altered, undelegations already in flight are only affected by at most the distance to the following newly-defined epoch, rather than being multiplied out into the future or suddenly made available all at once in the present. This is the least-surprising behavior in case we decide to change these parameters (however we probably won't!).